# Cryptococcus_Kwoniella_CompGenomics
Cryptococcus and Kwoniella Comparative Genomic Analysis

## Introduction
These scripts and data support the comparative genomics analysis of Cryptococcus and Kwoniella, as published in Coelho & David-Palma et al., 2024.

## Scripts and Data
This repository contains Python and R scripts for statistical analysis and visualization, and tree files, facilitating replication of the figures and supplementary figures listed below:

1. Fig1_and_S1A_Species_phylogeny
2. Fig5A_Centromere_length_comparison
3. Fig5B_TE_Classes
4. Fig5C_Cen_and_Non-cen_LTRs
5. Fig5D_TEs_boxplot_Cryptococcus_vs_Kwoniella
6. Fig5E_TEs_boxplot_Cryptococcus_pathogen_vs_saprophytic
7. Fig5F_TEs_vs_Genome_size
8. Fig5G_TEs_vs_AvgCenLength
9. Fig7A_UpSet_plot_Cryptococcus_Kwoniella
10. Fig7B_Categorizing_orthology_classes
11. Fig8_D-lactate_dehydrogenase_phylogeny
12. S1FigD_GCcontent_ridgline_plot
13. S1FigsH-N_Genomic_features
14. S14FigC_Argonaute_phylogeny
15. S15FigC_Znf3_phylogeny
16. S16FigsC-H_Mini_chrs_plots

## Dependencies
This project requires R and/or Python 3.6+ and the packages indicated in each script.

## Installation
No specific installation is required. Simply clone this repository using `git clone`.

## Citation
For citing this work, please refer to:
Coelho & David-Palma et al., 2024 (under revision)

## Authors
Marco A. Coelho (Duke University) @ Heitman Lab
